### PR TITLE
[X86][CodeGen] Add base atan2 intrinsic lowering (p4)

### PR DESCRIPTION
**Pull request by tex3d**
This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Based on example PR #96222 and fix PR #101268, with some differences due to 2-arg intrinsic and intermediate refactor (RuntimeLibCalls.cpp).

- Add llvm.experimental.constrained.atan2 - Intrinsics.td, ConstrainedOps.def, LangRef.rst
- Add to ISDOpcodes.h and TargetSelectionDAG.td, connect to intrinsic in BasicTTIImpl.h, and LibFunc_ in SelectionDAGBuilder.cpp
- Update LegalizeDAG.cpp, LegalizeFloatTypes.cpp, LegalizeVectorOps.cpp, and LegalizeVectorTypes.cpp
- Update isKnownNeverNaN in SelectionDAG.cpp
- Update SelectionDAGDumper.cpp
- Update libcalls - RuntimeLibcalls.def, RuntimeLibcalls.cpp
- TargetLoweringBase.cpp - Expand for vectors, promote f16
- X86ISelLowering.cpp - Expand f80, promote f32 to f64 for MSVC

Part 4 for Implement the atan2 HLSL Function #70096.

### CODE REVIEW

**Review by RKSimon — COMMENTED** [review]
GlobalISel test coverage?

**Comment by tex3d** [comment]
> GlobalISel test coverage?

Some coverage was added in part 2 earlier [here](https://github.com/llvm/llvm-project/pull/110037/files#diff-1836dbff33b24748f65272b0993967995dea458fc25e36b21a67b305716c0cc5) when global op was added for SPIR-V, but there aren't dedicated MIR tests for SPIR-V.  Since x86_64 doesn't use GlobalISel, we will be testing that a lot more in forthcoming part 5 that adds aarch64 after this merges.

**Comment by bogner** [comment]
> > GlobalISel test coverage?
> 
> Some coverage was added in part 2 earlier [here](https://github.com/llvm/llvm-project/pull/110037/files#diff-1836dbff33b24748f65272b0993967995dea458fc25e36b21a67b305716c0cc5) when global op was added for SPIR-V, but there aren't dedicated MIR tests for SPIR-V. Since x86_64 doesn't use GlobalISel, we will be testing that a lot more in forthcoming part 5 that adds aarch64 after this merges.

X86 does have some GlobalISel implemented, but it doesn't look like any handling of trig functions is in place yet. I think we should probably implement the generic parts of this change along with a backend where we can actually test it (like AArch64), and simplify this PR to only be the X86 parts. Since none of the trig is implemented in X86 GlobalISel at this time, we can probably leave that aspect alone.

**Comment by tex3d** [comment]
Switching to draft to sort this out.  I will either try to land the aarch64 part first, or if that's too difficult, I'll strip the GlobalISel changes from this one.

**Comment by tex3d** [comment]
I've moved GlobalISel changes to the next aarch64 step.
@RKSimon I hope this satisfies the GlobalISel test concerns for this PR.

I gave up trying to rearrange the steps as the aarch64 change has common SelectionDAG (and other) dependencies on changes in this PR, so aarch64 lowering is still be staged after this one.

File: llvm/test/CodeGen/X86/fp-intrinsics.ll [hunk 1]
```diff
@@ -1,4 +1,4 @@
-; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --function fatan2 --version 2
```
Comment (farzonl) [inline llvm/test/CodeGen/X86/fp-intrinsics.ll:1]: Did we use the right arguments for this test check generation? Don’t think this line should be here.

**Review by farzonl — APPROVED** [review]

**Review by RKSimon — COMMENTED** [review]
please can you rebase? I've added the missing declarations to the strict libcall tests to make this PR atan2 specific

**Comment by tex3d** [comment]
> please can you rebase? I've added the missing declarations to the strict libcall tests to make this PR atan2 specific

Rebased.

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Pull request merged**